### PR TITLE
Feature/2309 language selector

### PR DESCRIPTION
**Pull request by tonyr59h**
Should resolve https://github.com/wordpress-mobile/WordPress-Android/issues/2309

Adds a Language setting that allows the user to select a locale to apply to the app.

### CODE REVIEW

**Comment by maxme** [comment]
@tonyr59h thanks for this. I think this is a tricky feature ;)

Easy fixes:
* +1 for having the `wp_` preferences key in `strings.xml` but we must mark them `translatable="false"`
* Selected language should be saved and restored when the app start (not sure this is so easy, I've read some bad things about this a long time ago).
* We should filter `availableLocales` and keep only languages we have translations for. Or only list languages we have translations for, see bullet point below.

The tricky point:
* I don't know if it's possible, but one idea behind #2309 was to be able to select languages unsupported by Android (Gaelic for instance).

**Comment by maxme** [comment]
Not sure about this one: forget language selection if the user signs out.

**Comment by tonyr59h** [comment]
> Selected language should be saved and restored when the app start (not sure this is so easy, I've read some bad things about this a long time ago).

Restarting the Activity after applying the new locale configuration seems to work without consequence.

> Not sure about this one: forget language selection if the user signs out.

I'm hesitant about this one because if a user intends to sign in again they will have to navigate back to language selection.

> I don't know if it's possible, but one idea behind #2309 was to be able to select languages unsupported by Android (Gaelic for instance).

Apparently [it's possible to](http://stackoverflow.com/questions/14823431/android-multi-and-minority-language-support-for-languages-with-no-locale#answers) create a locale for any language. We'd have to provide all the required resources.

**Comment by maxme** [comment]
> I'm hesitant about this one because if a user intends to sign in again they will have to navigate back to language selection.

Yep. I fear someone change the language, signs out, and then can't login and get stuck on a foreign language, I assume that would be pretty rare and the logins fields are pretty common.

> Apparently it's possible to create a locale for any language. We'd have to provide all the required resources.
> 
> EDIT: We can provide resources for any language and add it to the language list, just need to provide a strings file for it. Here's a fake language, zz.

Great, there is no unsupported language fully translated today, but I'll notify translators so they can work on it (at least a Gaelic translator wanted to work on it). When our first "unsupported language" arrive, we'll open a new issue and add it to that list.

**Comment by maxme** [comment]
In the Settings screen, we should move the `Language` button from the `About` section. Create a new section? (This screen is a mess :/ - will be cleaner with the "My sites" / "Me" separation in the future).

**Comment by tonyr59h** [comment]
> Create a new section?
:scream_cat: I think there are too many sections already and Language fits well enough in the About section.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SettingsFragment.java [hunk 1]
```diff
@@ -365,15 +376,64 @@ public boolean onPreferenceClick(Preference preference) {
...
+            listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+                @Override
+                public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                    Resources res = getResources();
+                    DisplayMetrics dm = res.getDisplayMetrics();
+                    Configuration conf = res.getConfiguration();
+                    conf.locale = new Locale(values[position]);
+                    res.updateConfiguration(conf, dm);
+
+                    mSettings.edit().putString(SETTINGS_PREFERENCES, availableLocales[position]).apply();
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SettingsFragment.java]: Off by one error here: `position` can be `availableLocales.length + 1` (from line 391) `availableLocales[availableLocales.length + 1]  will crash`

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -859,4 +881,39 @@
     <string name="hs__conversation_header">Support chat</string>
     <string name="hs__username_blank_error">Enter a valid name</string>
     <string name="hs__invalid_email_error">Enter a valid email address</string>
+
+    <!-- Available languages -->
+    <string-array name="available_languages" translatable="false">
```
Comment (maxme) [inline WordPress/src/main/res/values/strings.xml]: Remove `available_languages` from strings.xml now it's generated (and add `available_languages.xml` to the git repo).

**Comment by maxme** [comment]
* We should add "English US (en)" which is the default `values/strings.xml` to the list.
* Sort the list view by locale name
* Some locale names aren't displayed: zh-cn or en-gb in the following screenshot for instance:

screen shot 2015-03-09 at 16 37 59

**Comment by maxme** [comment]
> Some locale names aren't displayed: zh-cn or en-gb in the following screenshot for instance:

Seems to affect only language code with a country code (ie xx-yy)

**Comment by tonyr59h** [comment]
Using an RTL language seems to format properly.

@maxme do you still notice https://github.com/wordpress-mobile/WordPress-Android/issues/2278 if you select an RTL language (Hebrew). If not maybe we can close out that issue (and PR). I've tested 4.1.2 and 5.0.0 devices.

**Comment by maxme** [comment]
> Using an RTL language seems to format properly.
> 
> @maxme do you still notice #2278 if you select an RTL language (Hebrew). If not maybe we can close out that issue (and PR). I've tested 4.1.2 and 5.0.0 devices.

RTL languages always worked in the app but they work without `android:supportsRtl` enabled. So it's not a "true rtl support" (for instance the menu drawer opens on the right of the screen with true rtl support).

**Comment by maxme** [comment]
Selecting Chinese or Portuguese doesn't work (country code problem?)

**Comment by tonyr59h** [comment]
> Selecting Chinese or Portuguese doesn't work (country code problem?)

I was using the wrong constructor when a region was provided.

**Comment by maxme** [comment]
:shipit:

**Pull request merged**